### PR TITLE
[ASan][Windows] Synchronizing ASAN init on Windows

### PR DESCRIPTION
**Pull request by zacklj89**
This is an attempt to synchronize ASAN initialization using `__sanitizer::atomic`s on Windows. One case in particular is a C# application that loads an ASan instrumented library.

This addresses:
 
1. [This check failure](https://github.com/llvm/llvm-project/blob/1df5ea29b43690b6622db2cad7b745607ca4de6a/compiler-rt/lib/asan/asan_interceptors.h#L29C36-L29C36) if a function is intercepted on T1 and called on T2 before T1 is done with ASAN initialization.
2. [This check failure](https://github.com/llvm/llvm-project/blob/1df5ea29b43690b6622db2cad7b745607ca4de6a/compiler-rt/lib/asan/asan_rtl.cpp#L387) can happen in parallel when two threads call `ENSURE_ASAN_INITED`.
3. [This check failure](https://github.com/llvm/llvm-project/blob/1df5ea29b43690b6622db2cad7b745607ca4de6a/compiler-rt/lib/asan/asan_rtl.cpp#L481) when thread is created before the main thread gets a chance to get created due to when `CreateThread` is intercepted.

Currently, everything is blocked by `#if SANITIZER_WINDOWS` because I'm not 100% sure about how ASan is loaded on other platforms and the implications.

### CODE REVIEW

**Review by vitalybuka — CHANGES_REQUESTED** [review]
We like small commits here. Unfortunately github is not very suitable, or LLVM didn't figured our what is going to be a flow.

Still I will ask you extract into NFC patches:
1. NFC: All clang-format changes
2. NFC: Introduction of setters/getter, as-is without atomics
3. Real patch with windows synchronization

**Comment by zacklj89** [comment]
Thanks @vitalybuka! I was trying to add reviewers as well (you, @strega-nil , and @barcharcraz), but I was unable to. Are there specific permissions needed, or something I missed in the Contributing.md?

**Comment by vitalybuka** [comment]
> 1. NFC: All clang-format changes
> 2. NFC: Introduction of setters/getter, as-is without atomics
> 3. Real patch with windows synchronization

I mean separate pull request, so we can bisect issues if necessary.

File: compiler-rt/lib/asan/asan_rtl.cpp
Comment (vitalybuka) [inline compiler-rt/lib/asan/asan_rtl.cpp:78]: static

Comment (vitalybuka) [reply compiler-rt/lib/asan/asan_rtl.cpp:78]: No need to change now. I will push them separately myself.

File: compiler-rt/lib/asan/asan_rtl.cpp
Comment (vitalybuka) [inline compiler-rt/lib/asan/asan_rtl.cpp:74]: do we need to #if ?
maybe atomic everywhere?

File: compiler-rt/lib/asan/asan_rtl.cpp
Comment (vitalybuka) [inline compiler-rt/lib/asan/asan_rtl.cpp:117]: do we need some kind of yeld here?

File: compiler-rt/lib/asan/asan_thread.cpp
Comment (vitalybuka) [inline compiler-rt/lib/asan/asan_thread.cpp:100]: how newthread can happen? Isn't there is interceptor for CreateThread which should be executed first on main thread?

Comment (vitalybuka) [reply compiler-rt/lib/asan/asan_thread.cpp:100]: Looks like dead lock prone solution
Whoever is creator of this thread may wait to join this what, and prevent initialization happening

File: compiler-rt/lib/asan/asan_rtl.cpp
Comment (zacklj89) [reply compiler-rt/lib/asan/asan_rtl.cpp:74]: I wasn't sure how the change would be viewed, so I added this for Windows only. If it seems benign, I'll remove the `#if SANITIZER_WINDOWS`

File: compiler-rt/lib/asan/asan_thread.cpp
Comment (zacklj89) [reply compiler-rt/lib/asan/asan_thread.cpp:100]: From .NET, if a worker thread loads an ASAN instrumented binary, we can see all of the checks in the PR description fail based on timing.

For the check `((!asan_init_is_running && "ASan init calls itself!")) != (0)`, this can happen when one thread is initializing ASan and has already intercepted `malloc` or `Rtl*` APIs. While it's still initializing, another thread calls the intercepted API, which results in the `__asan::Allocator` making another call to `AsanInitInternal`, and the check fires.

For the main thread creation, this happens less frequently since the call to `CreateMainThread` is nearly at the end of `AsanInitInternal` shortly after `CreateThread` has been intercepted. After it's intercepted, the thread registry can attempt to create another thread before the call to create the main thread. I'm not sure how ASan can be dynamically loaded on other platforms besides Windows, so this might not be an issue outside of Windows.

File: compiler-rt/lib/asan/asan_rtl.cpp
Comment (zacklj89) [reply compiler-rt/lib/asan/asan_rtl.cpp:117]: Good point. I was hesitant to add yields or mutex usage, but I think yield makes sense here.

File: compiler-rt/lib/asan/asan_thread.cpp
Comment (zacklj89) [reply compiler-rt/lib/asan/asan_thread.cpp:100]: @vitalybuka, do the changes look sufficient to you after the latest push?

Comment (vitalybuka) [reply compiler-rt/lib/asan/asan_thread.cpp:100]: You replied as "will do" to some items, as #if
But I see update in this PR.

Also there are https://github.com/zahiraam/llvm-project/commit/8e9ce62cf037767963ecd1a48aeb491bc0ef5863
you want to create PR for them?

BTW. There is https://github.com/getcord/spr which is proposed to be used for stacked code review, similar to we have in Phabricator, you may try that.

**Comment by zacklj89** [comment]
run check-asan-dynamic

File: compiler-rt/lib/asan/asan_thread.cpp
Comment (vitalybuka) [reply compiler-rt/lib/asan/asan_thread.cpp:100]: I guess  just create PRs for 8ef594b and 8e9ce62
I will merge them asap, and you can rebase this one

Comment (zacklj89) [reply compiler-rt/lib/asan/asan_thread.cpp:100]: Apologies... I think I'm confused on how PRs work here. 😅 

Are you saying rebase this branch with those, and then merge? In the future, is there a document for me to reference as to not trouble you with this?

Comment (vitalybuka) [reply compiler-rt/lib/asan/asan_thread.cpp:100]: I we don't have documentation yet, after migrating to pull request.

So easiest way is to create sevaral branches in your fork and create PR from each one:
1. format
2. format + asan_init_is_running
3. format + asan_init_is_running + the rest  (this one)

2 and 3 will have include some code from PRs above, but they will go away after merging earlier one and rebase

https://github.com/getcord/spr does this stuff for you, but it will required undocumented setup

Comment (zacklj89) [reply compiler-rt/lib/asan/asan_thread.cpp:100]: @vitalybuka is this what you had in mind? I tried using spr

https://github.com/llvm/llvm-project/pull/73819
https://github.com/llvm/llvm-project/pull/73822
https://github.com/llvm/llvm-project/pull/73824

Comment (vitalybuka) [reply compiler-rt/lib/asan/asan_thread.cpp:100]: It's close, but I assume you have branch per PR locally, because PRs include diffs from other PRs

With SPR you can have all patches in local branch, and upload them with spr diff --all, then it will create stacked review similar we had with Phabricator.

Comment (vitalybuka) [reply compiler-rt/lib/asan/asan_thread.cpp:100]: And SPR is suppose to upload into upstream repo, not your fork
`llvm-project/.git/config` should look like:

```
	githubRemoteName = origin
	githubRepository = llvm/llvm-project
	githubMasterBranch = main
	branchPrefix = users/vitalybuka/spr/
	requireTestPlan = false
```

Comment (zacklj89) [reply compiler-rt/lib/asan/asan_thread.cpp:100]: I assumed I had to use my fork to create the PR since I don't have push permissions to llvm. I tried changing the configuration and received invalid PR.

I'll see if @barcharcraz can help me with this live today.

Comment (vitalybuka) [reply compiler-rt/lib/asan/asan_thread.cpp:100]: Sorry for that, I expected it would be easier.
We can go one by one with regular PRs.

Comment (zacklj89) [reply compiler-rt/lib/asan/asan_thread.cpp:100]: I was able to get LLVM push permissions, and I created the PRs using `spr diff --all` on the main repo. Here is the last stacked PR. https://github.com/llvm/llvm-project/pull/74086

**Review by vitalybuka — CHANGES_REQUESTED** [review]
Could you please rebase the PR so we could continue the review?

**Comment by zacklj89** [comment]
Apologies @vitalybuka, just updated it.

The comment on the other PR about moving to `asan_win.cpp` seems ok, but I think that having all thread related code in `asan_thread.cpp` is fine too.

I will be out of office until the new year, so if this doesn't get merged I'll check back in then. Happy holidays!

**Comment by zacklj89** [comment]
@vitalybuka @barcharcraz is there anything further needed here?

**Comment by zacklj89** [comment]
I will close this for now since there has been no activity, and I believe @barcharcraz is in the process of making changes.

**Pull request closed without merging**
